File: breadth_first_search.js
const tap = require("tap");

function search(graph, name, check) {
  const q = [name];
  const searched = {};
  let person;

  while (q.length) {
    person = q.shift();

    if (!searched[person]) {
      for (let friend of graph[person]) {
        if (check(friend)) {
          return true;
        }

        searched[person] = true;
        q.push(...graph[person]);
      }
    }
  }

  return false;
}

const graph = {};
graph.you = ["alice", "bob", "claire"];
graph.bob = ["anuj", "peggy"];
graph.alice = ["peggy"];
graph.claire = ["thom", "jonny"];
graph.anuj = [];
graph.peggy = [];
graph.thom = [];
graph.jonny = [];

function check(str) {
  return str[str.length - 1] === "m";
}

tap.ok(search(graph, "you", check));

File: breadth_first_search.go
package grok

func search(graph map[string][]string, name string, check func(string) bool) bool {
	var searchQueue []string
	searchQueue = append(searchQueue, graph[name]...)
	searched := make(map[string]bool)
	var person string

	for len(searchQueue) != 0 {
		person = searchQueue[0]
		searchQueue = searchQueue[1:]

		if !searched[person] {
			if check(person) {
				return true
			}

			searchQueue = append(searchQueue, graph[person]...)
			searched[person] = true
		}
	}
	return false
}
